ENH: Add two private ufuncs related to generalized harmonic numbers.

WarrenWeckesser opened the pull request:
Two private ufuncs are added to `scipy.special._ufuncs`:

* `_gen_harmonic(n, a)` computes the generalized harmonic number

$$\sum_{i=1}^n i^{-a}$$

* `_normalized_gen_harmonic(j, k, n, a)` computes
  
$$\frac{\displaystyle\sum_{i=j}^k i^{-a}}{\displaystyle\sum_{i=1}^n i^{-a}}$$

These functions are used in `scipy.stats.zipfian` to improve performance and to improve precision when a is greater than but close to 1.

These aren't made part of the public API, so I don't think they need to become part of `xsf`.  That could be done in the future if there is downstream interest in such functions.

#### Reference issue

This change was inspired by https://github.com/scipy/scipy/issues/23487, but it doesn't close that issue.  The change makes the generation of random variates in `zipfian.rvs()` significantly faster, but there is another change (to be submitted in a follow-up PR) that should definitively resolve the slowness issue.

#### What does this implement/fix?

Before this change, the calculation of the harmonic numbers used in `stats.zipfian` when a <= 1 was implemented with a Python for-loop, so it was very slow.  This was especially evident in the generation of random variates, as noted in gh-23487.

This PR also improves the precision of the calculations in `zipfian` when `a` is greater than but close to 1.  When `a` > 1, the sum of the powers can be written $\zeta(a, m) - \zeta(a, n + 1)$, where $\zeta(a, m)$ is the Hurwitz zeta function.  Because of the subtraction, this formula loses numerical precision when $\zeta(a, m) \approx \zeta(a, n + 1)$.  In that case, the new function falls back to the simple sum of the powers. (See the comments in the code for more precise details.)

Peiffap commented on scipy/stats/_discrete_distns.py:
```diff
@@ -5,7 +5,8 @@
 from scipy import special
-from scipy.special import entr, logsumexp, betaln, gammaln as gamln, zeta
+from scipy.special import entr, logsumexp, betaln, gammaln as gamln
+from scipy.special._ufuncs import _gen_harmonic
```
The rest of the file seems to use imports as `scu.<fn>`.

WarrenWeckesser replied:
Good point! I'll fix that to be consistent with the other uses.

WarrenWeckesser replied:
Done.

WarrenWeckesser commented:
The inefficiency of the Zipfian CDF and SF calculations,

    CDF:  scu._gen_harmonic(1, k, a) / scu._gen_harmonic(1, n, a)
    SF:   scu._gen_harmonic(k + 1, n, a) / scu._gen_harmonic(1, n, a)

irks me.  A normalized generalized harmonic function would make each of those a single function call, in which the partial sum is computed in the same loop as the total sum.  As long as I'm adding private ufuncs, I might as well handle that too.

I'm moving the PR to draft while I see how well that works out.

WarrenWeckesser committed:
ENH: Add two private ufuncs related to generalized harmonic numbers.

Two private ufuncs are added to `scipy.special._ufuncs`:

* '_gen_harmonic(n, a)' computes the generalized harmonic number:

        sum_{i=1}^n i**-a

* '_normalized_gen_harmonic(j, k, n, a)' computes

        sum_{i=j}^k i**-a
        -----------------
        sum_{i=1}^n i**-a

These functions are used in scipy.stats.zipfian to improve
performance and to improve precision when a is greater than
but close to 1.

Peiffap commented on scipy/special/gen_harmonic.h:
```diff
@@ -0,0 +1,231 @@
+                 "n >= 1 is required, but got n = %" PRId64, n);
+        return NAN;
+    }
+    if (std::isnan(a)) {
+        if (n == 1) {
+            // IEEE: pow(1.0, nan) is 1.0.
+            return 1.0;
+        }
+        return NAN;
+    }
+    if (std::isinf(a)) {
+        if (a > 0) {
+            // a = +inf
+            return 1.0;
+        }
+        else {
+            // a = -inf
+            if (n == 1) {
+                // IEEE: pow(1.0, inf) = 1
+                return 1.0;
+            }
+            else {
+                return INFINITY;
+            }
+        }
+    }
```
Is there a reason not to do this? Same can be done in the normalized version.
```suggestion
    if (n == 1) {
        // IEEE: pow(1.0, _) is 1.0.
        return 1.0;
    }
    if (std::isnan(a)) {
        return NAN;
    }
    if (std::isinf(a)) {
        if (a > 0) {
            // a = +inf
            return 1.0;
        }
        else {
            // a = -inf
            return INFINITY;
        }
    }
```

Peiffap commented on scipy/special/gen_harmonic.h:
```diff
@@ -0,0 +1,231 @@
+        return sum_powers(1, n, a);
+    }
+    else {
+        // If here, we know a is finite and a > 1, and n >= 1.
```
This would need to change too then.
```suggestion
        // If here, we know a is finite and a > 1, and n > 1.
```

Peiffap reviewed (commented):
A quick question...

WarrenWeckesser committed:
Rearrange handling of special cases in gen_harmonic().

Co-authored-by: Gilles Peiffer <gilles.peiffer.yt@gmail.com>

WarrenWeckesser committed:
Rearrange handling of special cases in normalized_gen_harmonic().

WarrenWeckesser committed:
Use numeric_limits instead of NAN and INFINITY macros.

WarrenWeckesser committed:
Be less conservative about switching from the zeta formula to summing the powers.

WarrenWeckesser committed:
Fix typo in comment.

WarrenWeckesser committed:
Remove _gen_harmonic and _normalized_gen_harmonic from functions.json.

steppi reviewed (approved):
Thanks @WarrenWeckesser. I think the implementation looks solid. I left one comment on some things I think are worth thinking about regarding `dtypes`, targeted for the general audience (so may sound like I'm trying to explain things that I'm actually confident you know already). 

These functions will probably find their way into `xsf` for the sake of CuPy support, but it's definitely easier for now to add them here. I'd like to set things up so that contributing to `xsf` is painless, but there's more work needed to get to that point.
